File: rb101_programming_foundations/rb109_general_ruby_assessment/exercises/101-109_review/easy/letter_counter_pt1.rb
# Source: https://launchschool.com/exercises/e66bd409

# Write a method that takes a string with one or more space separated words
# and returns a hash that shows the number of words of different sizes.

def word_sizes(str)
  return_hsh = {}
  word_size_arr = str.split.map(&:size)
  word_size_arr.uniq.each do |len|
    return_hsh[len] = word_size_arr.count(len)
  end
  return_hsh
end

p word_sizes('Four score and seven.') == { 3 => 1, 4 => 1, 5 => 1, 6 => 1 }
p word_sizes('Hey diddle diddle, the cat and the fiddle!') == { 3 => 5, 6 => 1, 7 => 2 }
p word_sizes("What's up doc?") == { 6 => 1, 2 => 1, 4 => 1 }
p word_sizes('') == {}

File: exercises/210-javascript_smallproblems/easy_3/letter_counter_pt1.js
// Write a function that takes a string consisting of one or more space separated words,
// and returns an object that shows the number of words of different sizes.
// Words consist of any sequence of non-space characters.

function wordSizes(phrase) {
  if (!phrase) return {};

  let words = phrase.split(' ');
  return words.map(word => word.length)
              .reduce((countObj, length) => {
                countObj[length] = (countObj[length] || 0) + 1;
                return countObj;
              }, {});
}

console.log(wordSizes('Four score and seven.'));                       // { "3": 1, "4": 1, "5": 1, "6": 1 }
console.log(wordSizes('Hey diddle diddle, the cat and the fiddle!'));  // { "3": 5, "6": 1, "7": 2 }
console.log(wordSizes("What's up doc?"));                              // { "2": 1, "4": 1, "6": 1 }
console.log(wordSizes(''));                                            // {}